File: code/backend/core/src/test/java/games/paths/core/service/EchoServiceTest.java
package games.paths.core.service;

import games.paths.core.port.EchoPort;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for {@link EchoService}.
 * Verifies server status reporting, timestamp accuracy, and property management.
 */
@ExtendWith(MockitoExtension.class)
class EchoServiceTest {

    private EchoPort echoPort;
    private static final String DEFAULT_STATUS = "OK";
    private static final Map<String, String> DEFAULT_PROPS = Map.of(
            "env", "test",
            "version", "1.0.0-TEST",
            "applicationName", "paths-game-test"
    );

    @BeforeEach
    void setUp() {
        // Inizializzazione del servizio con proprietà standard di test
        echoPort = new EchoService(DEFAULT_STATUS, DEFAULT_PROPS);
    }

    // --- SEZIONE: STATO DEL SERVER ---

    @Nested
    @DisplayName("Server Status Tests")
    class StatusTests {

        @Test
        @DisplayName("Should return the status configured during initialization")
        void getServerStatus_returnsConfiguredStatus() {
            assertEquals(DEFAULT_STATUS, echoPort.getServerStatus());
        }

        @Test
        @DisplayName("Should return custom status when initialized with different values")
        void getServerStatus_returnsCustomStatus() {
            EchoPort maintenanceService = new EchoService("MAINTENANCE", Map.of());
            assertEquals("MAINTENANCE", maintenanceService.getServerStatus());
        }
    }

    // --- SEZIONE: GESTIONE PROPRIETÀ ---

    @Nested
    @DisplayName("Server Properties Tests")
    class PropertiesTests {

        @Test
        @DisplayName("Should provide access to all configured server properties")
        void getServerProperties_returnsAllProperties() {
            Map<String, String> props = echoPort.getServerProperties();
            
            assertAll("Property mapping validation",
                () -> assertEquals("test", props.get("env")),
                () -> assertEquals("1.0.0-TEST", props.get("version")),
                () -> assertEquals("paths-game-test", props.get("applicationName")),
                () -> assertEquals(3, props.size())
            );
        }

        @Test
        @DisplayName("Should return an unmodifiable map to protect internal state")
        void getServerProperties_isImmutable() {
            Map<String, String> props = echoPort.getServerProperties();
            // Verifica che il tentativo di modifica lanci l'eccezione corretta
            assertThrows(UnsupportedOperationException.class, () -> props.put("unauthorized", "write"));
        }

        @Test
        @DisplayName("Should handle cases with no properties configured")
        void getServerProperties_handlesEmptyMap() {
            EchoPort minimalService = new EchoService(DEFAULT_STATUS, Map.of());
            assertNotNull(minimalService.getServerProperties());
            assertTrue(minimalService.getServerProperties().isEmpty());
        }
    }

    // --- SEZIONE: TIMESTAMP E DIAGNOSTICA ---

    @Nested
    @DisplayName("Diagnostics Tests")
    class DiagnosticTests {

        @Test
        @DisplayName("Should return a timestamp within the execution time window")
        void getTimestamp_isApproximatelyCurrent() {
            // Branch: Verifica che il timestamp generato sia coerente con il clock di sistema
            long before = System.currentTimeMillis();
            long timestamp = echoPort.getTimestamp();
            long after = System.currentTimeMillis();

            assertTrue(timestamp >= before && timestamp <= after, 
                "Timestamp should be between " + before + " and " + after);
        }
    }
}

File: code/backend-php/tests/Unit/Core/Service/EchoServiceTest.php
<?php

declare(strict_types=1);

namespace Games\Paths\Tests\Unit\Core\Service;

use Games\Paths\Core\Service\EchoService;
use PHPUnit\Framework\TestCase;

class EchoServiceTest extends TestCase
{
    private EchoService $service;

    protected function setUp(): void
    {
        $this->service = new EchoService();
    }

    public function testGetStatusReturnsOk(): void
    {
        $status = $this->service->getStatus();
        $this->assertSame('OK', $status['status']);
    }

    public function testGetStatusContainsTimestamp(): void
    {
        $before = time() * 1000;
        $status = $this->service->getStatus();
        $after = time() * 1000;

        $this->assertArrayHasKey('timestamp', $status);
        $this->assertGreaterThanOrEqual($before, $status['timestamp']);
        $this->assertLessThanOrEqual($after, $status['timestamp']);
    }

    public function testGetStatusContainsProperties(): void
    {
        $status = $this->service->getStatus();

        $this->assertArrayHasKey('properties', $status);
        $this->assertArrayHasKey('env', $status['properties']);
        $this->assertArrayHasKey('version', $status['properties']);
        $this->assertArrayHasKey('applicationName', $status['properties']);
        $this->assertArrayHasKey('phpVersion', $status['properties']);
    }

    public function testApplicationNameIsCorrect(): void
    {
        $status = $this->service->getStatus();
        $this->assertSame('paths-game-backend-php', $status['properties']['applicationName']);
    }

    public function testPhpVersionIsSet(): void
    {
        $status = $this->service->getStatus();
        $this->assertSame(PHP_VERSION, $status['properties']['phpVersion']);
    }

    public function testNoJavaVersionKey(): void
    {
        $status = $this->service->getStatus();
        $this->assertArrayNotHasKey('javaVersion', $status['properties']);
    }
}
